### DISCUSSION TITLE
How to calculate timediff of two devices

### DISCUSSION THREAD

**Discussion by baobao2016** [lf-edge/ekuiper#1789] [Use Case]
应用场景 ：
1.当状态变化时，获取这个状态持续时间，
       比如当前机器是run状态,当状态发生变化时，我需要判断run状态持续了多久时间
       SELECT datediff( second, run status time, run status stop time) from demo
       请问这样的业务在以后的新版中是否可以增加支持,工业物联网中这个是常用到基本数据（用于计算OEE）

2.产量发生变化时，当前的产量与上一次产量之前的差值
     请问这样的业务在SQl中怎么写

3.当状态不发生变化时，是否可以查询间隔多少秒的数据（实现数据没有变化时，定时上传）
  主要是满足以下两个场景：3.1 数据有变化时触发（这个已实现）
                                              3.2 数据无变化定时触发，如每隔60秒上传一次，减少上传到云端的数据量（不知如何实现）

4.sql方式复杂的业务比较难实现，是否可以考虑规则里可以使用脚本来实现过滤数据

**Comment by ngjaying** [comment]
1. 我再想想，可能需要对当前的分析函数继续优化
2. `SELECT amount - lag(amount) as diff FROM demo where diff != 0`
3. 整两个场景需要合并成一个规则吗？单独做 3.2的话，可以用 `SELECT collect(*)[-1] FROM demo GROUP BY TumblingWindow(ss,60)`
4. 当前需要将脚本包装成自定义函数，然后在 SQL 里使用。新版本有考虑在 Graph API/Flow editor 里添加脚本节点，直接inline写脚本，可关注这个issue https://github.com/lf-edge/ekuiper/issues/1463

**Comment by baobao2016** [comment]
1.当状态变化时，获取这个状态持续时间，数据格式如下
timediff ( 2021-01-01 06:43:00.337, 2021-01-07 11:45:15.633)
最终形成这样的数据：Equipcode,StatusDesc,StartTime,EndTime,TimeDiff（The unit is second）
![image](https://user-images.githubusercontent.com/17892202/199420232-7fa8bd62-9094-4a64-ad30-e5aafd461896.png)

2.单独3.2，如每60秒上传一次数据，sink时只上传最后一条数据

**Comment by ngjaying** [comment]
好的，目前 1 我能想到的办法是给[状态分析函数](https://ekuiper.org/docs/zh/latest/sqls/built-in_functions.html#%E5%88%86%E6%9E%90%E5%87%BD%E6%95%B0)增加条件判断，从而可以对比某种条件，例如本例里的状态变化条件下的状态
`SELECT lag(StatusDesc) as status, StartTime - lag(StartTime) WHEN had_changed(true, StatusCode), EquipCode FROM demo WHERE had_changed(true, StatusCode)`
这里面的where是在状态变化的时候触发，select里选了变化之前的状态，变化后的StartTime（新状态开始时间，假设为旧状态结束时间） - 上一次变化发生时的 StartTime （即旧状态开始时间）
上一次变化发生时的 StartTime 用到了尚未支持的 WHEN 语法。如果这样能满足需求的话，我们可以考虑 WHEN的支持。

**Comment by baobao2016** [comment]
我想起用类似sql语法 case when..的方式来，select case when had_changed(true, StatusCode)==true Then CurrentSystemTime-lag(StartTime)  when had_changed(true, StatusCode)==false   Then XXXXX  End as timediff from demo ,类似这个意思，lag(xxx)就是取的最近的上一条数据的字段值吧

**Comment by baobao2016** [comment]
我还有一个需求，我计算好的数据能临时保存吗，等下次计算时，临时数据再跟流数据比较，进行计算，现在原始数据用lag能取到，比较我现在进行了一次计算，保存到内存表时，当我需要时再拿出来使用，请问目前有这个功能吗。

**Comment by ngjaying** [comment]
> 我想起用类似sql语法 case when..的方式来，select case when had_changed(true, StatusCode)==true Then CurrentSystemTime-lag(StartTime) when had_changed(true, StatusCode)==false Then XXXXX End as timediff from demo ,类似这个意思，lag(xxx)就是取的最近的上一条数据的字段值吧

这个应该不行，lag作为状态函数每次都会计算，不是满足过滤条件才计算。所以才需要添加 WHEN 的语法。

> 我还有一个需求，我计算好的数据能临时保存吗，等下次计算时，临时数据再跟流数据比较，进行计算，现在原始数据用lag能取到，比较我现在进行了一次计算，保存到内存表时，当我需要时再拿出来使用，请问目前有这个功能吗。

这个目前应该是支持的，lag的参数可以是任何表达式

**Comment by baobao2016** [comment]
请问保存计算好的临时数据，应该怎么实现呢。

**Comment by baobao2016** [comment]
ekuiper 用在云端时，是一台设备创建一个流，还是多台设备共用一个流，在云端会处理多台设备，这样计算时数据流入是乱序的，我怎么针对每台设备计算呢，比较状态变化时，获取的是这台设备上一次的最新数据，一台设备一个流能规避这个问题，只是设备多了，要创建好多流行，比较几百台设备时，如果共用一个流，是不是可以用partition by分组设备，分组后相当于每台设备单独计算，类似多线程单独处理每台设备的数据。不知我的这个理解正确不？

**Comment by ngjaying** [comment]
> 请问保存计算好的临时数据，应该怎么实现呢。

lag(a+b)

> ekuiper 用在云端时，是一台设备创建一个流，还是多台设备共用一个流

都可以

> 如果共用一个流，是不是可以用partition by分组设备

有状态的分析函数可以用 partition by

**Comment by baobao2016** [comment]
我想把流入的数据其中的一个字段做为查询条件，比如根据 设备编码或设备ID查询它的lag，然后新旧数据进行计算，这个可能只创建一个共用的流就行，维护进来比较简单。请问ekuiper有QQ或微信群吗。

**Comment by ngjaying** [comment]
> 请问ekuiper有QQ或微信群吗。

https://ekuiper.org/docs/zh/latest/#%E8%8E%B7%E5%8F%96%E5%B8%AE%E5%8A%A9
